src/training_algorithms/sgd_training_algorithm.cpp

```diff
@@ -0,0 +1,278 @@
+    model.reset_mode(c, execution_mode::training);
+    model.reset_epoch_statistics(execution_mode::training);
+    do_epoch_begin_cbs(model);
+
+    // Training iterations
+    if (num_batches > 0) {
+      for (size_t i = 0; i < num_batches; i++) { train_mini_batch(c, model); }
+    } else {
+      while (!train_mini_batch(c, model)) {}
+    }
+
+    // Finalize epoch
+    c.inc_epoch();
+    model.reconcile_weight_values();
+    do_epoch_end_cbs(model);
+
+    // Evaluate on validation set
+    auto key = c.get_trainer().check_and_build_execution_context(c, model, execution_mode::validation);
+    auto& evaluation_context = static_cast<sgd_execution_context&>(c.get_trainer().get_execution_context(key));
+    evaluate(evaluation_context, model, execution_mode::validation);
+  }
+  do_train_end_cbs(model);
+}
+
+////////////////////////////////////////////////////////////
+// Evaluation and training
+////////////////////////////////////////////////////////////
+
+bool sgd_training_algorithm::train_mini_batch(sgd_execution_context& c,
+                                              model& model) {
+  model.reset_mode(c, execution_mode::training);
+  do_batch_begin_cbs(model, execution_mode::training);
+
+  bool finished;
+
+#if defined(LBANN_HAVE_OMP_TASKLOOP)
+  LBANN_OMP_PARALLEL
+  {
+    #pragma omp single
+    {
+#endif
+  // Forward prop step
+  model.clear_gradients();
+  model.forward_prop(execution_mode::training);
+  // Result is not needed until the end of the mini-batch.
+  model.get_objective_function()->start_evaluation(execution_mode::training,
+                                                    c.get_current_mini_batch_size());
+
+  // Backward prop step
+  model.get_objective_function()->differentiate();
+  model.backward_prop();
+  model.get_objective_function()->compute_weight_regularization();
+
+  // Finish evaluation.
+  model.get_objective_function()->finish_evaluation(execution_mode::training,
+                                                     c.get_current_mini_batch_size());
+  model.evaluate_metrics(execution_mode::training,
+                          c.get_current_mini_batch_size());
+
+  // Update step
+  model.update_weights();
+  finished = model.update_layers();
+#if defined(LBANN_HAVE_OMP_TASKLOOP)
+    }
+  }
+#endif
+
+  c.inc_step();
+  do_batch_end_cbs(model, execution_mode::training);
+  return finished;
+}
+
+void sgd_training_algorithm::evaluate(sgd_execution_context& c,
+                                      model& model,
+                                      execution_mode mode,
+                                      size_t num_batches) {
+  // Return early if execution mode is invalid
+  if (!model.is_execution_mode_valid(mode)) return;
+  if (mode != execution_mode::validation
+      && mode != execution_mode::testing) {
+    std::stringstream err;
+    err << __FILE__ << " " << __LINE__ << " :: "
+        << "invalid execution mode for evaluation";
+    throw lbann_exception(err.str());
```

Comment:
Delete all the stuff in this block and replace with: 
```suggestion
    LBANN_ERROR("invalid execution mode for evaluation");
```